projects/update_success_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,383 @@
+
+## Make sure that the volume of incoming pings is reasonable
+Check if the volume of `update` pings with `reason = ready` matches with the volume of pings with `reason = success`. For each ping with `reason = ready`, find the matching ping with `reason = success`.
+
+We are considering the data within a very narrow window of time: we could see `reason = success` pings from users that sent a `reason = ready` ping before the 3rd of September and `reason = ready` pings from users that have sent us a `reason = success` after the 9th of September. Filter these edge cases out by inspecting the `previousBuildId` and `targetBuildId`.
+
+
+```python
+filtered_ready = ping_ready.filter(lambda p: p.get("payload/targetBuildId") < "{}000000".format(MAX_DATE))
+filtered_success = ping_success.filter(lambda p: p.get("payload/previousBuildId") >= "{}999999".format(MIN_DATE))
+```
+Use the filtered RDDs to match betweek the different ping reasons.
+
+
+```python
+# Get an identifier that keeps in consideration both the current build
+# and the target build.
+ready_uuid = filtered_ready\
+    .map(lambda p: (p.get("clientId"),
+                    p.get("application/buildId"),
+                    p.get("application/channel"),
+                    p.get("application/version"),
+                    p.get("payload/targetBuildId"),
+                    p.get("payload/targetChannel"),
+                    p.get("payload/targetVersion")))
+
+# Get an identifier that considers both the prevous build info and the
+# current build info. The order of the values in the tuple need to match
+# the one from the previous RDD.
+success_uuid = filtered_success\
+    .map(lambda p: (p.get("clientId"),
+                    p.get("payload/previousBuildId"),
+                    p.get("payload/previousChannel"),
+                    p.get("payload/previousVersion"),
+                    p.get("application/buildId"),
+                    p.get("application/channel"),
+                    p.get("application/version")))
+```
+Let's match each `reason = ready` ping with a `reason = success` one, and count them.
+
+
+```python
+matching_update_pings = ready_uuid.intersection(success_uuid)
+matching_update_ping_count = matching_update_pings.count()
+```
+Finally, show up some stats.
+
+
+```python
+print("{:.3f}% of the 'update' ping with reason 'ready' have a matching ping with reason 'success'."\
+      .format(pct(matching_update_ping_count, filtered_ready.count())))
+print("{:.3f}% of the 'update' ping with reason 'success' have a matching ping with reason 'ready'."\
+      .format(pct(matching_update_ping_count, filtered_success.count())))
+```
+    60.383% of the 'update' ping with reason 'ready' have a matching ping with reason 'success'.
```

Comment:
This %ge went _down_ when we excluded reason=ready pings that were updating to a build we knew we didn't have a reason=success ping for? That seems odd.